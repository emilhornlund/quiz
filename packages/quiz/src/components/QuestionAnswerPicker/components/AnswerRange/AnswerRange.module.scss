@use '../../../../styles/colors';
@use '../../../../styles/helpers';
@use '../../../../styles/variables';

.main {
  display: flex;
  flex-direction: column;
  flex: 1;

  .interactive, .interactive > form {
    display: flex;
    flex-direction: column;
    justify-content: center;
    flex: 1;

    @include helpers.mobile {
      row-gap: variables.$mobile-layout-gutter;
    }

    @include helpers.tablet {
      row-gap: variables.$tablet-layout-gutter;
    }

    @include helpers.desktop {
      row-gap: variables.$desktop-layout-gutter;
    }

    input[type=range] {
      display: flex;
      flex-direction: column;
      width: 100%;
      padding: 0;
      margin: 0;
      appearance: none;
      -webkit-appearance: none;
      box-sizing: border-box;
      background-color: colors.$yellow-2;

      @include helpers.mobile {
        height: helpers.pxToRem(14px);
        border-radius: helpers.pxToRem(calc(14px / 2));
        -webkit-box-shadow: inset 0 0 0 helpers.pxToRem(2px) colors.$white;
        -moz-box-shadow: inset 0 0 0 helpers.pxToRem(2px) colors.$white;
        box-shadow: inset 0 0 0 helpers.pxToRem(2px) colors.$white;
      }

      @include helpers.tablet {
        height: helpers.pxToRem(18px);
        border-radius: helpers.pxToRem(calc(18px / 2));
        -webkit-box-shadow: inset 0 0 0 helpers.pxToRem(4px) colors.$white;
        -moz-box-shadow: inset 0 0 0 helpers.pxToRem(4px) colors.$white;
        box-shadow: inset 0 0 0 helpers.pxToRem(4px) colors.$white;
      }

      @include helpers.desktop {
        height: helpers.pxToRem(22px);
        border-radius: helpers.pxToRem(calc(22px / 2));
        -webkit-box-shadow: inset 0 0 0 helpers.pxToRem(6px) colors.$white;
        -moz-box-shadow: inset 0 0 0 helpers.pxToRem(6px) colors.$white;
        box-shadow: inset 0 0 0 helpers.pxToRem(6px) colors.$white;
      }

      &::-webkit-slider-thumb, &::-moz-range-thumb {
        -webkit-appearance: none;
        appearance: none;
        background: colors.$yellow-2;
        cursor: pointer;
        transition: 0.1s ease-in-out;
        border-style: solid;
        border-color: colors.$white;
        box-sizing: border-box;

        @include helpers.mobile {
          width: helpers.pxToRem(30px);
          height: helpers.pxToRem(30px);
          border-radius: helpers.pxToRem(15px);
          border-width: helpers.pxToRem(4px);
        }

        @include helpers.tablet {
          width: helpers.pxToRem(40px);
          height: helpers.pxToRem(40px);
          border-radius: helpers.pxToRem(20px);
          border-width: helpers.pxToRem(6px);
        }

        @include helpers.desktop {
          width: helpers.pxToRem(50px);
          height: helpers.pxToRem(50px);
          border-radius: helpers.pxToRem(25px);
          border-width: helpers.pxToRem(8px);
        }
      }
    }

    .inputGroup {
      display: flex;
      flex-direction: row;
      justify-content: center;
      align-items: center;

      @include helpers.mobile {
        column-gap: variables.$mobile-narrow-layout-gutter;
      }

      @include helpers.tablet {
        column-gap: variables.$tablet-narrow-layout-gutter;
      }

      @include helpers.desktop {
        column-gap: variables.$desktop-narrow-layout-gutter;
      }

      div:first-child {
        flex: 1;
      }
    }
  }

  .nonInteractive {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    background-color: colors.$turquoise-2;
    flex: 1;

    @include helpers.mobile {
      column-gap: variables.$mobile-layout-gutter;
      padding: variables.$mobile-layout-gutter;
      border-radius: helpers.pxToRem(8px);
    }

    @include helpers.tablet {
      column-gap: variables.$tablet-layout-gutter;
      padding: variables.$tablet-layout-gutter;
      border-radius: helpers.pxToRem(12px);
    }

    @include helpers.desktop {
      column-gap: variables.$desktop-layout-gutter;
      padding: variables.$desktop-layout-gutter;
      border-radius: helpers.pxToRem(16px);
    }

    svg {
      display: flex;
      flex-direction: column;

      @include helpers.mobile {
        @include helpers.fontSize(28px);
      }

      @include helpers.tablet {
        @include helpers.fontSize(38px);
      }

      @include helpers.desktop {
        @include helpers.fontSize(48px);
      }
    }

    span {
      display: flex;
      flex-direction: column;
      font-weight: bold;
      width: 100%;
      flex: 1;

      @include helpers.mobile {
        @include helpers.fontSize(18px);
      }

      @include helpers.tablet {
        @include helpers.fontSize(28px);
      }

      @include helpers.desktop {
        @include helpers.fontSize(38px);
      }
    }
  }
}
