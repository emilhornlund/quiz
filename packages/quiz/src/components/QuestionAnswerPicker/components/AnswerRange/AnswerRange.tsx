import React, { ChangeEvent, FC, FormEvent, useCallback, useState } from 'react'

import NonInteractiveInfoBox from '../../../../states/common/NonInteractiveInfoBox'
import { isValidNumber } from '../../../../utils/helpers.ts'
import Button from '../../../Button'
import TextField from '../../../TextField'

import styles from './AnswerRange.module.scss'

export interface AnswerRangeProps {
  min: number
  max: number
  step: number
  interactive: boolean
  onSubmit: (value: number) => void
}

const AnswerRange: FC<AnswerRangeProps> = ({
  min,
  max,
  step,
  interactive = true,
  onSubmit,
}) => {
  const [value, setValue] = useState<number>(Math.floor(min + (max - min) / 2))
  const [valid, setValid] = useState<boolean>(false)

  const handleSliderChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      const newValue = parseInt(event.target.value, 10)
      setValue(newValue)
      setValid(isValidNumber(newValue, min, max))
    },
    [min, max],
  )

  const handleSubmit = useCallback(
    (event: FormEvent) => {
      event.preventDefault()
      if (valid) {
        onSubmit(value)
      }
    },
    [onSubmit, valid, value],
  )

  return (
    <div className={styles.main}>
      {interactive ? (
        <div className={styles.interactive}>
          <form onSubmit={handleSubmit}>
            <input
              id="slider-range"
              type="range"
              min={min}
              max={max}
              step={step}
              value={value}
              onChange={handleSliderChange}
            />
            <div className={styles.inputGroup}>
              <TextField
                id="slider-input"
                type="number"
                min={min}
                max={max}
                value={value}
                onChange={(newValue) => setValue(newValue as number)}
                onValid={setValid}
                required
              />
              <Button
                id="submit-button"
                type="submit"
                value="Submit"
                disabled={!valid}
              />
            </div>
          </form>
        </div>
      ) : (
        <>
          <div style={{ display: 'flex', flex: 1 }} />
          <NonInteractiveInfoBox
            info={`Pick an answer between ${min} and ${max} on your screen`}
          />
        </>
      )}
    </div>
  )
}

export default AnswerRange
