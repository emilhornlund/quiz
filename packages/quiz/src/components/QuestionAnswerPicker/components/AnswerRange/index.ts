export type { AnswerRangeProps } from './AnswerRange'
export { default } from './AnswerRange'
