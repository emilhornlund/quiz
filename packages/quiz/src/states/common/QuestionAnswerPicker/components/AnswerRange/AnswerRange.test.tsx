import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { describe, expect, it, vi } from 'vitest'

import AnswerRange from './AnswerRange'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
vi.mock('../../../../../components', async (orig: any) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const actual: any = await orig()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const TextField = ({ id, value, onChange, onValid, min, max }: any) => (
    <div>
      <input
        data-testid="tf-input"
        id={id}
        type="number"
        min={min}
        max={max}
        value={String(value ?? '')}
        onChange={(e) =>
          onChange?.(Number((e.target as HTMLInputElement).value))
        }
      />
      <button
        type="button"
        data-testid="tf-valid"
        onClick={() => onValid?.(true)}>
        make-valid
      </button>
      <button
        type="button"
        data-testid="tf-invalid"
        onClick={() => onValid?.(false)}>
        make-invalid
      </button>
    </div>
  )
  return { ...actual, TextField }
})

describe('AnswerRange', () => {
  it('renders interactive mode with initial midpoint and disabled submit', () => {
    const { container } = render(
      <AnswerRange min={0} max={100} step={5} interactive onSubmit={vi.fn()} />,
    )
    const slider = screen.getByRole('slider') as HTMLInputElement
    expect(slider.value).toBe('50')
    const submit = screen.getByRole('button', { name: /submit/i })
    expect(submit).toBeDisabled()
    expect(container).toMatchSnapshot()
  })

  it('enables submit when slider value is valid and submits that value', () => {
    const onSubmit = vi.fn()
    render(
      <AnswerRange
        min={10}
        max={60}
        step={5}
        interactive
        onSubmit={onSubmit}
      />,
    )

    const slider = screen.getByRole('slider') as HTMLInputElement
    fireEvent.change(slider, { target: { value: '25' } })

    const submit = screen.getByRole('button', { name: /submit/i })
    expect(submit).toBeEnabled()

    fireEvent.click(submit)
    expect(onSubmit).toHaveBeenCalledTimes(1)
    expect(onSubmit).toHaveBeenCalledWith(25)
  })

  it('updates via TextField, toggles validity, and submits new value', () => {
    const onSubmit = vi.fn()
    render(
      <AnswerRange min={0} max={10} step={1} interactive onSubmit={onSubmit} />,
    )

    fireEvent.click(screen.getByTestId('tf-valid'))

    const tf = screen.getByTestId('tf-input')
    fireEvent.change(tf, { target: { value: '7' } })

    const submit = screen.getByRole('button', { name: /submit/i })
    expect(submit).toBeEnabled()

    fireEvent.click(submit)
    expect(onSubmit).toHaveBeenCalledTimes(1)
    expect(onSubmit).toHaveBeenLastCalledWith(7)
  })

  it('reflects slider attributes from props', () => {
    render(
      <AnswerRange min={3} max={33} step={3} interactive onSubmit={vi.fn()} />,
    )
    const slider = screen.getByRole('slider') as HTMLInputElement
    expect(slider).toHaveAttribute('min', '3')
    expect(slider).toHaveAttribute('max', '33')
    expect(slider).toHaveAttribute('step', '3')
  })

  it('renders non-interactive mode with info message and no form', () => {
    const { container } = render(
      <AnswerRange
        min={5}
        max={15}
        step={1}
        interactive={false}
        onSubmit={vi.fn()}
      />,
    )
    expect(
      screen.getByText(/Pick an answer between 5 and 15 on your screen/i),
    ).toBeInTheDocument()
    expect(screen.queryByRole('slider')).not.toBeInTheDocument()
    expect(
      screen.queryByRole('button', { name: /submit/i }),
    ).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
