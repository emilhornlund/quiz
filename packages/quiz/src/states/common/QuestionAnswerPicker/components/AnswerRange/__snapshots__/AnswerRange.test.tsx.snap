// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AnswerRange > renders interactive mode with initial midpoint and disabled submit 1`] = `
<div>
  <div
    class="answerRange"
  >
    <div
      class="interactive"
    >
      <form>
        <input
          id="slider-range"
          max="100"
          min="0"
          step="5"
          type="range"
          value="50"
        />
        <div
          class="inputGroup"
        >
          <div>
            <input
              data-testid="tf-input"
              id="slider-input"
              max="100"
              min="0"
              type="number"
              value="50"
            />
            <button
              data-testid="tf-valid"
              type="button"
            >
              make-valid
            </button>
            <button
              data-testid="tf-invalid"
              type="button"
            >
              make-invalid
            </button>
          </div>
          <div
            class="buttonInputContainer buttonInputKindPrimary"
          >
            <button
              data-testid="test-submit-button-button"
              disabled=""
              id="submit-button"
              name="submit-button"
              type="submit"
            >
              <span>
                Submit
              </span>
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`AnswerRange > renders non-interactive mode with info message and no form 1`] = `
<div>
  <div
    class="answerRange"
  >
    <div
      style="display: flex; flex: 1 1 0%;"
    />
    <div
      class="nonInteractiveInfoBox"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-info"
        data-icon="circle-info"
        data-prefix="fas"
        role="img"
        viewBox="0 0 512 512"
      >
        <path
          d="M256 512a256 256 0 1 0 0-512 256 256 0 1 0 0 512zM224 160a32 32 0 1 1 64 0 32 32 0 1 1 -64 0zm-8 64l48 0c13.3 0 24 10.7 24 24l0 88 8 0c13.3 0 24 10.7 24 24s-10.7 24-24 24l-80 0c-13.3 0-24-10.7-24-24s10.7-24 24-24l24 0 0-64-24 0c-13.3 0-24-10.7-24-24s10.7-24 24-24z"
          fill="currentColor"
        />
      </svg>
      <span>
        Pick an answer between 5 and 15 on your screen
      </span>
    </div>
  </div>
</div>
`;
